Temporary secure memory for TA2TA

jenswi-linaro opened the pull request:
Extracted the early "building blocks" of https://github.com/OP-TEE/optee_os/pull/3036 which still makes sense on their own with the final commit "libutee: allocate temp secmem for invoke"

jenswi-linaro commented:
With `core: user_ta: keep load segs in linked list` fixed for a memory leak in `load_elf_from_store()`

jenswi-linaro commented:
It seems that the lockdep stuff doesn't work well with dynamically allocated and freed mutexes, which is used in the commit "￼core: file: refactor interface to be thread safe".

jenswi-linaro commented:
Adds "libutils: sys/queue.h: add STAILQ_REMOVE_AFTER()" and "core: lockdep: support lock destroy" which should be put first in this PR before merging. "core: lockdep: support lock destroy" takes care of mutexes which are destroyed as in the commit "core: file: refactor interface to be thread safe".

etienne-lms commented on core/include/mm/file.h:
```diff
@@ -26,17 +27,43 @@ struct file_slice {
+void file_lock(struct file *f);
+
+/*
+ * file_lock() - Tries to lock the file without blocking
```
minor: Tries or Try

etienne-lms commented on core/include/mm/file.h:
```diff
@@ -26,17 +27,43 @@ struct file_slice {
+void file_unlock(struct file *f);
+
+/*
+ * file_add_slice() - add a slice to a file
```
minor
s/add/Add/
maybe few indentation to align below

etienne-lms commented on core/arch/arm/mm/tee_mmu.c:
```diff
@@ -317,40 +317,86 @@ TEE_Result vm_map(struct user_ta_ctx *utc, vaddr_t *va, size_t len,
+
+	if (mobj_is_paged(r->mobj)) {
+		tee_pager_rem_uta_region(utc, r->va, r->size);
+		maybe_free_pgt(utc, r);
```
`maybe_free_pgt()` needs to be called for unpaged regions at other places.

jenswi-linaro replied:
I'm not sure I understand. You mean I should move this call outside the block to always be called regardless if it's paged or not?

etienne-lms replied:
I think so, yes, as in `tee_mmu_rem_rwmem()` and `tee_mmu_clean_param()`. Am I wrong?

jenswi-linaro replied:
You're right, well spotted.

etienne-lms commented on core/arch/arm/mm/tee_mmu.c:
```diff
@@ -352,19 +355,25 @@ TEE_Result vm_set_prot(struct user_ta_ctx *utc, vaddr_t va, size_t len,
 	if ((r->attr & TEE_MATTR_PROT_MASK) == prot)
 		return TEE_SUCCESS;
 
+	was_writeable = (r->attr & (TEE_MATTR_UW | TEE_MATTR_PW));
```
maybe drop outer parentheses.

etienne-lms commented on core/arch/arm/mm/tee_mmu.c:
```diff
@@ -352,19 +355,25 @@ TEE_Result vm_set_prot(struct user_ta_ctx *utc, vaddr_t va, size_t len,
 
-	r->attr &= ~TEE_MATTR_PROT_MASK;
-	r->attr |= prot & TEE_MATTR_PROT_MASK;
+
```
may remove empty lines

etienne-lms commented on core/include/mm/tee_mmu.h:
```diff
@@ -23,9 +23,29 @@ void vm_info_final(struct user_ta_ctx *utc);
  * Desired virtual address can be specified in @va otherwise @va must be
  * initialized to 0 if the next available can be chosen.
+ *
+ * @pad_begin and @pad_end specifies how much extra free space should be
```
s/specifies/specify/

etienne-lms commented on core/include/mm/tee_mmu.h:
```diff
@@ -23,9 +23,29 @@ void vm_info_final(struct user_ta_ctx *utc);
+ *
+ * @pad_begin and @pad_end specifies how much extra free space should be
+ * kept when establishing the map. This allows mapping the first part of
+ * for instance and ELF file while knowing that the next part which has to
```
a/and ELF/an ELF/

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -981,26 +1007,22 @@ static TEE_Result load_elf_from_store(const TEE_UUID *uuid,
-	 * relocation is updated.
-	 */
-	res = register_ro_slices(utc, elf_state, segs, num_segs);
+
```
remove empty line

etienne-lms commented on core/arch/arm/include/kernel/user_ta.h:
```diff
@@ -94,6 +97,35 @@ static inline TEE_Result tee_ta_init_user_ta_session(
+}
+#endif
+
+
```
remove empty line.

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -1272,3 +1271,62 @@ TEE_Result tee_ta_init_user_ta_session(const TEE_UUID *uuid,
+		goto err;
+
+	seg->va = *va;
+	seg->size = f->num_pages * SMALL_PAGE_SIZE;
```
`seg->size` already set.

jenswi-linaro commented:
Comments addressed

jenswi-linaro commented:
The checkpatch warning is just a complaint that the commit description of one of the "review" patches is a bit short.

jenswi-linaro commented:
Squashed fixups and applied tags. Moved  "libutils: sys/queue.h: add STAILQ_REMOVE_AFTER()", "core: lockdep: support lock destroy" and "Add SLIST_HEAD() and SLIST_ENTRY() to typedefs.checkpatch" first in the patch list.

jbech-linaro commented:
> Squashed fixups and applied tags.

After checking the patches, there are still many commits with no tags. If not all commits have tags, then please emphasize that, just so we avoid pushing things by mistake (having that said, my "merge routine" is to check this before merging).

etienne-lms commented on lib/libutee/arch/arm/user_ta_entry.c:
```diff
@@ -119,6 +119,67 @@ static void ta_header_remove_session(uint32_t session_id)
+static void to_utee_params(struct utee_params *up, uint32_t param_types,
+			   const TEE_Param params[TEE_NUM_PARAMS])
+{
+	size_t n;
```
minor: init local vars

jenswi-linaro replied:
I'll fix

etienne-lms commented on lib/libutee/include/tee_internal_api_extensions.h:
```diff
@@ -34,4 +34,25 @@ TEE_Result TEE_CacheClean(char *buf, size_t len);
 TEE_Result TEE_CacheInvalidate(char *buf, size_t len);
 
+/*
+ * tee_map_zi() - Map zero inititialized memory
```
s/inititialized/initialized/

jenswi-linaro replied:
I'll fix

etienne-lms commented on lib/libutee/tee_system_pta.c:
```diff
@@ -0,0 +1,79 @@
+					       TEE_PARAM_TYPE_VALUE_INOUT,
+					       TEE_PARAM_TYPE_VALUE_INPUT,
+					       TEE_PARAM_TYPE_NONE);
+	TEE_Param params[TEE_NUM_PARAMS];
```
`= { };`

jenswi-linaro replied:
I'll fix

etienne-lms commented on lib/libutee/tee_system_pta.c:
```diff
@@ -0,0 +1,79 @@
+	TEE_Param params[TEE_NUM_PARAMS];
+	TEE_Result res = TEE_SUCCESS;
+
+	memset(params, 0, sizeof(params));
```
s/memset()/TEE_MemFill()/?
Or remove since initialized where defined.

jenswi-linaro replied:
I'll remove it.

etienne-lms commented on lib/libutee/tee_api.c:
```diff
@@ -16,36 +16,102 @@ static const void *tee_api_instance_data;
+			 * invoke_command/open_session below.
+			 */
+			if (b && s &&
+			    !TEE_CheckMemoryAccessRights(flags, b, s)) {
```
minor: would prefer `if (b && s && TEE_CheckMemoryAccessRights(flags, b, s) == TEE_SUCCESS)` which is more explicit than `if (b && s && !TEE_CheckMemoryAccessRights(flags, b, s))`.

jenswi-linaro replied:
This is a matter of taste. The two expressions are equivalent and both styles are OK with our coding style. The style used here is normally the preferred style so I'd rather keep it as it is.

etienne-lms commented on core/include/kernel/lockdep.h:
```diff
@@ -103,6 +103,12 @@ static inline void lockdep_lock_release(struct lockdep_lock_head *owned,
+ */
+void lockdep_lock_destroy(struct lockdep_node_head *graph, uintptr_t id);
+
+
```
remove empty line

jenswi-linaro replied:
I'll fix

etienne-lms commented on core/include/kernel/lockdep.h:
```diff
@@ -103,6 +103,12 @@ static inline void lockdep_lock_release(struct lockdep_lock_head *owned,
 }
 
+/*
+ * Destroy lock @id in @graph. The lock is freed
```
minor: terminal dot.

jenswi-linaro replied:
I'll fix

jenswi-linaro commented:
> If not all commits have tags, then please emphasize that

@jbech-linaro, fair enough. I'll keep that in mind.

jenswi-linaro commented on core/arch/arm/include/kernel/user_ta.h:
```diff
@@ -125,6 +125,18 @@ static inline TEE_Result user_ta_unmap(struct user_ta_ctx *utc __unused,
+TEE_Result user_ta_set_prot(struct user_ta_ctx *utc, vaddr_t va, size_t len,
+			    uint32_t prot);
+#else
+static inline TEE_Result user_ta_unmap(struct user_ta_ctx *utc __unused,
```
I'll fix

jenswi-linaro commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -1330,3 +1330,29 @@ TEE_Result user_ta_unmap(struct user_ta_ctx *utc, vaddr_t va, size_t len)
+	if (!seg)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (seg->file && (prot & (TEE_MATTR_UW | TEE_MATTR_PW)))
```
I'll fix

jenswi-linaro commented:
Applied tags together squashing in trivial fixes to the review comments
where applicable.

Tag for "libutee: allocate temp secmem for invoke" is not applied
as we're still discussing that point.

Note that the following commits still lacks tags:
"libutee: allocate temp secmem for invoke"
"libutee: support dynamic memory mapping"
"core: add user_ta_set_prot()"

jenswi-linaro commented:
The IBART build error is due to xtest now needing `PTA_SYSTEM_DERIVE_TA_UNIQUE_KEY`.
I'll rebase to fix the error.

jenswi-linaro commented:
Why is shippable disabled?

jenswi-linaro commented:
Rebased on master. There was some conflicts in "core: pta_system: support dynamic memory mapping" which I had to resolve.

jenswi-linaro commented:
Added the commit "core: fix access flags in system_derive_ta_unique_key()" to fix the previous error in 4013.

jenswi-linaro commented:
How can we move forward with this? As noted above, some commits still needs a R-B or A-B. Is it easier if I squash in the fixups?

jforissier commented:
For commit "core: lockdep: support lock destroy":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented:
For commit "core: lockdep: support lock destroy":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented:
For commit "core: fix access flags in system_derive_ta_unique_key()":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Squashed, rebased and tags applied. Note that a few commits still miss R-B or A-B tags.

jforissier commented:
For commit "libutee: support dynamic memory mapping":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Applied tag. Now only two commits are missing R-B or A-B tags, namely:
- "core: add user_ta_set_prot()"
- "libutee: allocate temp secmem for invoke"

Can we make one final effort to clear this so we can get this out of the way for #3036?

jforissier commented:
For both commit "core: add user_ta_set_prot()" and ""libutee: allocate temp secmem for invoke": LGTM.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

But perhaps @etienne-lms wants to give an A-b/R-b too, since he has made some comments already.

jenswi-linaro commented:
Tags applied
